// Copyright (c) 2020 The Orbit Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

syntax = "proto3";

package orbit_grpc_protos;

import "module.proto";
import "tracepoint.proto";

// This is needed by LockFreeBufferCaptureEventProducer.
option cc_enable_arenas = true;

message CaptureOptions {
  bool trace_context_switches = 1;
  int32 pid = 2;
  double sampling_rate = 3;

  enum UnwindingMethod {
    kUndefined = 0;
    kFramePointers = 1;
    kDwarf = 2;
  }
  UnwindingMethod unwinding_method = 4;

  message InstrumentedFunction {
    string file_path = 1;
    uint64 file_offset = 2;
    uint64 function_id = 3;

    enum FunctionType {
      kRegular = 0;
      kTimerStart = 1;
      kTimerStop = 2;
    }
    FunctionType function_type = 4;
  }

  repeated InstrumentedFunction instrumented_functions = 5;

  bool trace_thread_state = 8;

  bool trace_gpu_driver = 6;

  repeated TracepointInfo instrumented_tracepoint = 7;

  bool enable_introspection = 9;

  uint64 max_local_marker_depth_per_command_buffer = 10;

  bool collect_memory_info = 11;
  uint64 memory_sampling_period_ns = 12;
}

// For CaptureEvents with a duration, excluding for now GPU-related ones, we
// represent the time interval as (duration_ns, end_timestamp_ns). As durations
// are smaller in magnitude than absolute timestamps, and as protobufs use
// variable-length encoding, this saves some bandwidth compared to
// (start_timestamp_ns, end_timestamp_ns).
// The pair (duration_ns, end_timestamp_ns) was preferred over
// (start_timestamp_ns, duration_ns) because this way, when perf_event_open
// events are processed in order, the CaptureEvents are sorted by the only
// absolute timestamp contained in them (as they are obviously generated when
// the "end" perf_event_open event is processed).

message SchedulingSlice {
  reserved 4;
  int32 pid = 1;
  int32 tid = 2;
  int32 core = 3;
  uint64 duration_ns = 6;
  uint64 out_timestamp_ns = 5;
}

message FunctionCall {
  reserved 4;
  int32 pid = 1;
  int32 tid = 2;
  uint64 function_id = 3;
  uint64 duration_ns = 9;
  uint64 end_timestamp_ns = 5;
  int32 depth = 6;
  uint64 return_value = 7;
  repeated uint64 registers = 8;
}

message IntrospectionScope {
  reserved 3;
  int32 pid = 1;
  int32 tid = 2;
  uint64 duration_ns = 7;
  uint64 end_timestamp_ns = 4;
  int32 depth = 5;
  repeated uint64 registers = 6;
}

message ApiEvent {
  uint64 num_raw_events = 1;
  bytes raw_data = 2;
}

message Callstack {
  repeated uint64 pcs = 1;
}

message InternedCallstack {
  uint64 key = 1;
  Callstack intern = 2;
}

message CallstackSample {
  int32 pid = 1;
  int32 tid = 2;
  uint64 callstack_id = 3;
  uint64 timestamp_ns = 4;
}

message FullCallstackSample {
  int32 pid = 1;
  int32 tid = 2;
  Callstack callstack = 3;
  uint64 timestamp_ns = 4;
}

message InternedString {
  uint64 key = 1;
  string intern = 2;
}

message InternedTracepointInfo {
  uint64 key = 1;
  TracepointInfo intern = 2;
}

message TracepointEvent {
  int32 pid = 1;
  int32 tid = 2;
  uint64 timestamp_ns = 3;
  int32 cpu = 4;
  uint64 tracepoint_info_key = 5;
}

message FullTracepointEvent {
  int32 pid = 1;
  int32 tid = 2;
  uint64 timestamp_ns = 3;
  int32 cpu = 4;
  TracepointInfo tracepoint_info = 5;
}

message FullGpuJob {
  int32 pid = 1;
  int32 tid = 2;
  uint32 context = 3;
  uint32 seqno = 4;
  int32 depth = 5;
  uint64 amdgpu_cs_ioctl_time_ns = 6;
  uint64 amdgpu_sched_run_job_time_ns = 7;
  uint64 gpu_hardware_start_time_ns = 8;
  uint64 dma_fence_signaled_time_ns = 9;
  string timeline = 10;
}

message GpuJob {
  int32 pid = 1;
  int32 tid = 2;
  uint32 context = 3;
  uint32 seqno = 4;
  int32 depth = 5;
  uint64 amdgpu_cs_ioctl_time_ns = 6;
  uint64 amdgpu_sched_run_job_time_ns = 7;
  uint64 gpu_hardware_start_time_ns = 8;
  uint64 dma_fence_signaled_time_ns = 9;
  uint64 timeline_key = 10;
}

message GpuQueueSubmission {
  GpuQueueSubmissionMetaInfo meta_info = 1;
  repeated GpuSubmitInfo submit_infos = 2;
  repeated GpuDebugMarker completed_markers = 3;
  // This is the total number of begin markers submitted in this submission.
  int32 num_begin_markers = 4;
}

message GpuQueueSubmissionMetaInfo {
  int32 tid = 1;
  int32 pid = 4;
  uint64 pre_submission_cpu_timestamp = 2;
  uint64 post_submission_cpu_timestamp = 3;
}

message GpuSubmitInfo {
  repeated GpuCommandBuffer command_buffers = 1;
}

message GpuCommandBuffer {
  uint64 begin_gpu_timestamp_ns = 1;
  uint64 end_gpu_timestamp_ns = 2;
}

message GpuDebugMarkerBeginInfo {
  // Begin markers can be submitted in a different submission then the end
  // markers, thus the submission meta info of a begin marker might differ from
  // the one of the submission containing the complete debug marker.
  GpuQueueSubmissionMetaInfo meta_info = 1;
  uint64 gpu_timestamp_ns = 2;
}

message GpuDebugMarker {
  GpuDebugMarkerBeginInfo begin_marker = 1;
  uint64 end_gpu_timestamp_ns = 2;
  uint64 text_key = 3;
  int32 depth = 4;
  Color color = 5;
}

message Color {
  // Color ranges are [0.f, 1.f] for the following fields.
  float red = 1;
  float green = 2;
  float blue = 3;
  float alpha = 4;
}

message ThreadName {
  int32 pid = 1;
  int32 tid = 2;
  string name = 3;
  uint64 timestamp_ns = 4;
}

message ThreadStateSlice {
  reserved 4;
  int32 pid = 1;  // pid is currently not set as we don't have the information.
  int32 tid = 2;

  // These are the ones listed in
  // /sys/kernel/debug/tracing/events/sched/sched_switch/format and in
  // https://github.com/torvalds/linux/blob/master/fs/proc/array.c in
  // task_state_array[], with the difference that the OS doesn't distinguish
  // between "running" (scheduled on a CPU) and "runnable" (ready to be
  // scheduled but not actually scheduled), rather they are the same state
  // (often referred to as "running", sometimes as "runnable or running").
  enum ThreadState {
    kRunning = 0;
    kRunnable = 1;
    kInterruptibleSleep = 2;
    kUninterruptibleSleep = 3;
    kStopped = 4;
    kTraced = 5;
    kDead = 6;
    kZombie = 7;
    kParked = 8;
    kIdle = 9;
  }
  ThreadState thread_state = 3;

  uint64 duration_ns = 6;
  uint64 end_timestamp_ns = 5;
}

message AddressInfo {
  uint64 absolute_address = 1;
  uint64 offset_in_function = 2;
  uint64 function_name_key = 3;
  uint64 module_name_key = 4;
}

message FullAddressInfo {
  uint64 absolute_address = 1;
  uint64 offset_in_function = 2;
  string function_name = 3;
  string module_name = 4;
}

message ModuleUpdateEvent {
  int32 pid = 1;
  uint64 timestamp_ns = 2;
  ModuleInfo module = 3;
}

message SystemMemoryUsage {
  uint64 timestamp_ns = 1;
  // These fields are retrieved from /proc/meminfo.
  int64 total_kb = 2;
  int64 free_kb = 3;
  int64 available_kb = 4;
  int64 buffers_kb = 5;
  int64 cached_kb = 6;
}

message ClientCaptureEvent {
  oneof event {
    // Note that field numbers from 1-15 take only 1 byte to encode
    // https://developers.google.com/protocol-buffers/docs/proto#assigning_field_numbers
    // use them for high frequency events. For the rest please assign
    // numbers starting with 16.
    //
    // Please keep these alphabetically ordered.

    // Even though AddressInfo is a high-frequency event
    // it is going to go away in the future when we switch to
    // frame-pointer based unwinding.
    AddressInfo address_info = 16;
    ApiEvent api_event = 24;
    CallstackSample callstack_sample = 1;
    FunctionCall function_call = 2;
    GpuJob gpu_job = 3;
    GpuQueueSubmission gpu_queue_submission = 4;
    InternedCallstack interned_callstack = 5;
    InternedString interned_string = 18;
    InternedTracepointInfo interned_tracepoint_info = 19;
    IntrospectionScope introspection_scope = 20;
    ModuleUpdateEvent module_update_event = 21;
    SchedulingSlice scheduling_slice = 6;
    SystemMemoryUsage system_memory_usage = 23;
    ThreadName thread_name = 22;
    ThreadStateSlice thread_state_slice = 7;
    TracepointEvent tracepoint_event = 8;
  }
}

message ProducerCaptureEvent {
  oneof event {
    // Note that field numbers from 1-15 take only 1 byte to encode
    // https://developers.google.com/protocol-buffers/docs/proto#assigning_field_numbers
    // use them for high frequency events. For the rest please assign
    // numbers starting with 16.
    //
    // Please keep these alphabetically ordered.
    ApiEvent api_event = 16;
    CallstackSample callstack_sample = 1;
    FullCallstackSample full_callstack_sample = 2;

    // Even though AddressInfo is a high-frequency event
    // it is going to go away in the future when we switch to
    // frame-pointer based unwinding.
    FullAddressInfo full_address_info = 16;
    FullGpuJob full_gpu_job = 3;
    FullTracepointEvent full_tracepoint_event = 4;
    FunctionCall function_call = 5;
    GpuQueueSubmission gpu_queue_submission = 6;
    InternedCallstack interned_callstack = 7;
    InternedString interned_string = 18;
    IntrospectionScope introspection_scope = 19;
    ModuleUpdateEvent module_update_event = 20;
    SchedulingSlice scheduling_slice = 8;
    SystemMemoryUsage system_memory_usage = 22;
    ThreadName thread_name = 21;
    ThreadStateSlice thread_state_slice = 9;
  }
}
